File: 322-coin-change/322-coin-change.rb
# @param {Integer[]} coins
# @param {Integer} amount
# @return {Integer}
def coin_change(coins, amount)
    dp = Array.new amount + 1, false
    dp[0] = 0
    (1..amount).each do |i|
        possibles = []
        coins.each do |coin|
            possibles << dp[i - coin] if i >= coin && dp[i - coin] 
        end
        dp[i] = possibles.min + 1 if !possibles.empty?
    end
    return dp[-1] ? dp[-1] : -1
end

File: 322-coin-change/322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */

// dynamic programming problem
// use tabulation/ dfs backtracking:
// bottom up on decision tree

// coinChange(coins, 0) === 0 
// can figure out the rest
const coinChange = function(coins, amount) {
    if (!amount) return amount;
    const dp = Array(amount + 1).fill(false);
    dp[0] = 0
    
    for (let i = 1; i <= amount; i++) {
        const poss = [];
        coins.forEach(c => {
            if (c <= i && typeof dp[i-c] === 'number') {
                poss.push(dp[i-c]);
            }
        })
        if (poss.length) dp[i] = Math.min(...poss) + 1;
    }
    return dp.at(-1) ? dp.at(-1) : -1;
};


